### DISCUSSION TITLE
Using DigitalOcean Object Storage instead of S3

### DISCUSSION THREAD

**Discussion by PhilReinking** [outline/outline#1372] [General] [answered]
Hey there,

Since I am not a AWS user, I tried to use the Digital Ocean object storage instead of S3.
I got it working by making some adjustments in the utils/s3.js

```
diff --git a/server/utils/s3.js b/server/utils/s3.js
index ee4245f4..4ec9d721 100644
--- a/server/utils/s3.js
+++ b/server/utils/s3.js
@@ -13,7 +13,7 @@ const AWS_REGION = process.env.AWS_REGION;
 const AWS_S3_UPLOAD_BUCKET_NAME = process.env.AWS_S3_UPLOAD_BUCKET_NAME;

 const s3 = new AWS.S3({
-  s3ForcePathStyle: true,
+  s3ForcePathStyle: false,
   accessKeyId: process.env.AWS_ACCESS_KEY_ID,
   secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
   endpoint: new AWS.Endpoint(process.env.AWS_S3_UPLOAD_BUCKET_URL),
@@ -84,9 +84,7 @@ export const publicS3Endpoint = (isServerUpload?: boolean) => {
     "localhost:"
   ).replace(/\/$/, "");

-  return `${host}/${isServerUpload && isDocker ? "s3/" : ""}${
-    process.env.AWS_S3_UPLOAD_BUCKET_NAME
-  }`;
+  return `${host}`;
 };
```

Is this something that I have to maintain myself, or should this be customizable through a setting? I'm not sure if the current implementation is only compatible with AWS S3 or if it does work with other S3 compatible storages like Minio? Does anybody have a good solution for this?

**Comment by tommoor** [comment, marked as answer]
> does work with other S3 compatible storages like Minio?

Yes, other self hosters have had success with this.

> should this be customizable through a setting?

I think we could modify this to work yea, that would be ideal. How about using `AWS_S3_UPLOAD_BUCKET_NAME` not being set then it follows similar logic to what you have above?

**Reply by PhilReinking** [reply]
I just tested to assign the `AWS_S3_UPLOAD_BUCKET_NAME` an empty value, but this is not possible without getting an error message when I try to start Outline.

`The AWS_S3_UPLOAD_BUCKET_NAME env variable must be set when using AWS`

I would suggest making the following changes:

- make the `AWS_S3_UPLOAD_BUCKET_NAME` optional (allow empty value)
- introduce a new `AWS_S3_FORCE_PATH_STYLE` setting (also optional, defaults to true for now, but should be set to false before September 30, 2020)

👉🏻 Amazon will deprecate the support for the path style model:

> Original Plan – Support for the path-style model ends on September 30, 2020.

> Revised Plan – Support for the path-style model continues for buckets created on or before September 30, 2020. Buckets created after that date must be referenced using the virtual-hosted model.

For Reference: https://aws.amazon.com/de/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story

Is there any reason why Outline uses these path style URLs, or is this something that could easily be changed?

**Reply by tommoor** [reply]
> introduce a new AWS_S3_FORCE_PATH_STYLE setting (also optional, defaults to true for now

I think instead of this we can just go ahead and change it over, we just use this style because the code is more than 3 years old and it hasn't been a priority to update as of yet. 

issue: https://github.com/outline/outline/issues/1375

**Comment by dickyj** [comment]
Hi,
Do you happen to have the full s3.js file for the digitalocean spaces? What do I need to fill in for AWS_REGION?

**Reply by PhilReinking** [reply]
the diff in my first post shows all changes I made.
As for the region, I filled that with ams3 in my configuration.
But I haven't checked if this is even necessary.
